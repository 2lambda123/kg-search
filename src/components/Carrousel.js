/*
*   Copyright (c) 2018, EPFL/Human Brain Project PCO
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*       http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
*/

import React, { PureComponent } from "react";
import { isMobile } from "../helpers/BrowserHelpers";
import "./Carrousel.css";


const CarrouselItem = ({data, showPrevious, onPrevious, onClose, itemComponent, navigationComponent}) => {
  const ItemComponent =  itemComponent;
  const NavigationComponent = navigationComponent;
  return (
    <div className="kgs-carrousel__panel">
      <div className="kgs-carrousel__header">
        {showPrevious && (
          <button className="kgs-carrousel__previous-button" onClick={onPrevious}>
            <i className="fa fa-chevron-left" /> Previous
          </button>
        )}
        <div className="kgs-carrousel__navigation">
          {data && (
            <NavigationComponent/>
          )}
        </div>
        <button className="kgs-carrousel__close-button" onClick={onClose}>
          <i className="fa fa-close" />
        </button>
      </div>
      <div className="kgs-carrousel__body">
        {data && (
          <ItemComponent data={data} />
        )}
      </div>
    </div>
  );
};

const nbOfViews = 5;

export class Carrousel extends PureComponent {
  constructor(props) {
    super(props);
    this.views =  Array.from(Array(nbOfViews)).map((x, i) => ({
      id: i + 1,
      data: null
    }));
  }
  componentDidMount() {
    if (!isMobile) {
      window.addEventListener("keyup", this._keyupHandler.bind(this), false);
    }
  }
  componentWillUnmount() {
    if (!isMobile) {
      window.removeEventListener("keyup", this._keyupHandler);
    }
  }
  _keyupHandler(event) {
    const {onPrevious, onClose} = this.props;
    if (this.props.show) {
      if (event.keyCode === 27) {
        event.preventDefault();
        typeof onClose === "function" && onClose();
      } else if (event.keyCode === 8) {
        event.preventDefault();
        typeof onPrevious === "function" && onPrevious();
      }
    }
  }
  render(){
    const {className, show, data, onPrevious, onClose, itemComponent, navigationComponent} = this.props;
    if (!show || !Array.isArray(data) || !data.length) {
      return null;
    }

    //window.console.debug("Carrousel rendering...", data);

    const selectedView = (data.length -1) % nbOfViews + 1;
    const views = this.views;
    views.forEach((view, idx) => {
      const idxData = (idx <= selectedView - 1)?data.length - (selectedView - idx):data.length - nbOfViews  +  (idx - selectedView);
      view.data = idxData >= 0?data[idxData]:null;
    });
    const showPrevious = data.length > 1;

    const classNames = ["kgs-carrousel", className].join(" ");

    return(
      <div data-selected={selectedView} className={classNames}>
        <div className="kgs-carrousel__views">
          {views.map(view => (
            <div key={view.id} className="kgs-carrousel__view" data-view={view.id} >
              <CarrouselItem data={view.data} showPrevious={showPrevious} onPrevious={onPrevious} onClose={onClose} itemComponent={itemComponent} navigationComponent={navigationComponent} />
            </div>
          ))}
        </div>
      </div>
    );
  }
}