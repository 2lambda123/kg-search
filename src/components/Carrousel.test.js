/*
*   Copyright (c) 2018, EPFL/Human Brain Project PCO
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*       http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
*/

import React from "react";
import renderer from "react-test-renderer";
import Enzyme, { mount, shallow, render } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Carrousel from "./Carrousel";

Enzyme.configure({ adapter: new Adapter() });

test('Carrousel component renders initially', () => {
    const component = renderer.create(
        <Carrousel className="className" show={true} value="a value" data={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} onPrevious={() => {}} onClose={() => {}} itemComponent={() => null} navigationComponent={() => null} cookielawBanner={() => null} noticeComponent={() => null} />
    );
  
    expect(component.toJSON()).toMatchSnapshot();
});

test('Carrousel test show false"', () => {
    const component = renderer.create(
        <Carrousel className="className" show={false} value="a value" data={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} onPrevious={() => {}} onClose={() => {}} itemComponent={() => null} navigationComponent={() => null} cookielawBanner={() => null} noticeComponent={() => null} />
    );
    expect(component.toJSON()).toBe(null);
});
  
test('Carrousel test className"', () => {
    const component = shallow(
        <Carrousel className="className" show={true} value="a value" data={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} onPrevious={() => {}} onClose={() => {}} itemComponent={() => null} navigationComponent={() => null} cookielawBanner={() => null} noticeComponent={() => null} />
    );
    expect(component.hasClass("className"));
});

test('Carrousel test number of items', () => {
    const component = render(
        <Carrousel className="className" show={true} value="a value" data={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} onPrevious={() => {}} onClose={() => {}} itemComponent={() => null} navigationComponent={() => null} cookielawBanner={() => null} noticeComponent={() => null} />
    );
    expect(component.find(".kgs-carrousel__item").length).toBe(5);
});

test('Carrousel test items', () => {
    const component = renderer.create(
        <Carrousel className="className" show={true} value="a value" data={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} onPrevious={() => {}} onClose={() => {}} itemComponent={() => null} navigationComponent={() => null} cookielawBanner={() => null} noticeComponent={() => null} />
    );
    const instance = component.getInstance();

    expect(instance.items.length).toBe(5);
    expect(instance.items[1].position).toBe(0);
    expect(instance.items[1].isActive);
    expect(instance.items[1].data.label).toBe("another label");
});

test('Carrousel test navigation buttons', () => {
    const previousFn = jest.fn();
    const closeFn = jest.fn();
    const component = mount(
        <Carrousel className="className" show={true} value="a value" data={[{label: "a label", value: "a value"},{label: "another label", value: "another value"}]} onPrevious={previousFn} onClose={closeFn} itemComponent={() => null} navigationComponent={() => null} cookielawBanner={() => null} noticeComponent={() => null} />
    );
    expect(component.find(".kgs-carrousel__previous-button").length).toBe(1);
    expect(component.find(".kgs-carrousel__close-button").length).toBe(1);
    component.find('.kgs-carrousel__previous-button').simulate('click');
    expect(previousFn.mock.calls.length).toBe(1);
    expect(closeFn.mock.calls.length).toBe(0);
    component.find('.kgs-carrousel__close-button').simulate('click');
    expect(previousFn.mock.calls.length).toBe(1);
    expect(closeFn.mock.calls.length).toBe(1);
});
