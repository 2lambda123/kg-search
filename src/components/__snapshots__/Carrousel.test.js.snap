// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carrousel component renders initially 1`] = `
<div
  className="kgs-carrousel className"
>
  <div
    className="kgs-carrousel__panel"
  >
    <div
      className="kgs-carrousel__item position4"
    >
      <div
        className="kgs-carrousel__content"
      >
        <div
          className="kgs-carrousel__header"
        >
          <div
            className="kgs-carrousel__navigation"
          />
        </div>
        <div
          className="kgs-carrousel__body"
        />
      </div>
    </div>
    <div
      className="kgs-carrousel__item position0"
    >
      <div
        className="kgs-carrousel__content"
      >
        <div
          className="kgs-carrousel__header"
        >
          <button
            className="kgs-carrousel__previous-button"
            onClick={[Function]}
          >
            <i
              className="fa fa-chevron-left"
            />
             Previous
          </button>
          <div
            className="kgs-carrousel__navigation"
          />
          <button
            className="kgs-carrousel__close-button"
            onClick={[Function]}
          >
            <i
              className="fa fa-close"
            />
          </button>
        </div>
        <div
          className="kgs-carrousel__body"
        />
      </div>
    </div>
    <div
      className="kgs-carrousel__item position1"
    >
      <div
        className="kgs-carrousel__content"
      >
        <div
          className="kgs-carrousel__header"
        >
          <div
            className="kgs-carrousel__navigation"
          />
        </div>
        <div
          className="kgs-carrousel__body"
        />
      </div>
    </div>
    <div
      className="kgs-carrousel__item position2"
    >
      <div
        className="kgs-carrousel__content"
      >
        <div
          className="kgs-carrousel__header"
        >
          <div
            className="kgs-carrousel__navigation"
          />
        </div>
        <div
          className="kgs-carrousel__body"
        />
      </div>
    </div>
    <div
      className="kgs-carrousel__item position3"
    >
      <div
        className="kgs-carrousel__content"
      >
        <div
          className="kgs-carrousel__header"
        >
          <div
            className="kgs-carrousel__navigation"
          />
        </div>
        <div
          className="kgs-carrousel__body"
        />
      </div>
    </div>
  </div>
</div>
`;
